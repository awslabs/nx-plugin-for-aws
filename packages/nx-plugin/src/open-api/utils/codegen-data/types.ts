/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { Plugin } from '@hey-api/openapi-ts';
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type ClientData = Parameters<Plugin.LegacyHandler<any>>[0]['client'];
export type Operation = ClientData['services'][number]['operations'][number];
export type Model = ClientData['models'][number];

export interface CodeGenData extends ClientData {
  className: string;
  info: OpenAPIV3.InfoObject | OpenAPIV3_1.InfoObject;
  allOperations: Operation[];
  operationsByTag: { [tag: string]: Operation[] };
  untaggedOperations: Operation[];
  vendorExtensions: { [key: string]: any };
}

export const flattenModelLink = (link?: Model | Model[]): Model =>
  link === undefined ? undefined : Array.isArray(link) ? link[0] : link;

// Model types which indicate it is composed (ie inherits/mixin's another schema)
export const COMPOSED_SCHEMA_TYPES = new Set(['one-of', 'any-of', 'all-of']);
export const COLLECTION_TYPES = new Set(['array', 'dictionary']);
export const PRIMITIVE_TYPES = new Set([
  'string',
  'integer',
  'number',
  'boolean',
  'null',
  'any',
  'binary',
  'void',
]);
