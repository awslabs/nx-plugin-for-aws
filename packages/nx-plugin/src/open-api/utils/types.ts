/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type Spec = OpenAPIV3.Document | OpenAPIV3_1.Document;
