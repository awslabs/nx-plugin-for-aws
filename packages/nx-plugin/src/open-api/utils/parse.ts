/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import SwaggerParser from '@apidevtools/swagger-parser';
import { Spec } from './types';

const TREE_PROTOCOL_PREFIX = 'workspace://';

/**
 * Parse an OpenAPI specification given its path within the tree
 */
export const parseOpenApiSpec = async (
  tree: Tree,
  specPath: string,
): Promise<Spec> => {
  // Use a custom protocol prefix to ensure that the swagger parser doesn't try and read from the filesystem itself
  const spec = await SwaggerParser.bundle(
    `${TREE_PROTOCOL_PREFIX}/${specPath}`,
    {
      resolve: {
        // Use a custom resolver which reads files from the tree rather than directly from the filesystem
        file: {
          order: 1,
          read: (file) => {
            const path = file.url.slice(TREE_PROTOCOL_PREFIX.length);
            if (!tree.exists(path)) {
              const message = `Unable to find ${specPath} in workspace ${tree.root}`;

              // Write an error to console as swagger parser's error message may be confusing with our custom protocol prefix
              console.error(message);
              throw new Error(message);
            }
            return tree.read(path);
          },
          canRead: (file) => file.url.startsWith(TREE_PROTOCOL_PREFIX),
        },
        http: false,
      },
    },
  );

  if ('swagger' in spec) {
    throw new Error(
      `OpenAPI v2 specifications are not supported. Please use OpenAPI v3`,
    );
  }
  return spec;
};
