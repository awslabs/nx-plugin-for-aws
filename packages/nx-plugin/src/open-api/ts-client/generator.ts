/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiTsClientGeneratorSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import * as path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';
import { formatFilesInSubtree } from '../../utils/format';

/**
 * Generates typescript client from an openapi spec
 */
export const openApiTsClientGenerator = async (
  tree: Tree,
  options: OpenApiTsClientGeneratorSchema,
) => {
  const data = await buildOpenApiCodeGenerationData(
    tree,
    options.openApiSpecPath,
  );

  generateOpenApiTsClient(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Build a data structure which can be used to generate code from OpenAPI
 */
export const buildOpenApiCodeGenerationData = async (
  tree: Tree,
  specPath: string,
) => {
  const spec = await parseOpenApiSpec(tree, specPath);
  return await buildOpenApiCodeGenData(spec);
};

/**
 * Generate an OpenAPI typescript client in the target directory
 */
export const generateOpenApiTsClient = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsClientGenerator;
