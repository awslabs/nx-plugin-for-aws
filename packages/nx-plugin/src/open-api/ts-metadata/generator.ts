/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiMetadataSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { formatFilesInSubtree } from '../../utils/format';
import path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';

/**
 * Generates typescript metadata from an openapi spec
 */
export const openApiTsMetadataGenerator = async (
  tree: Tree,
  options: OpenApiMetadataSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsMetadata(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Generate OpenAPI typescript metadata in the target directory
 */
export const generateOpenApiTsMetadata = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsMetadataGenerator;
