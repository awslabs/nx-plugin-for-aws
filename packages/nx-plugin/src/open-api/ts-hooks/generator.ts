/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { OpenApiHooksSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { generateOpenApiTsClient } from '../ts-client/generator';

/**
 * Generates typescript hooks from an openapi spec
 */
export const openApiTsHooksGenerator = async (
  tree: Tree,
  options: OpenApiHooksSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsClient(tree, data, options.outputPath);

  // TODO: generate hooks which wrap the client
};

export default openApiTsHooksGenerator;
