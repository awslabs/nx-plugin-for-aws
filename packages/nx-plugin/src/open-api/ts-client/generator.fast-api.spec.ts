/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { openApiTsClientGenerator } from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { Spec } from '../utils/types';
import { baseUrl, callGeneratedClient } from './generator.utils.spec';
import { Tree } from '@nx/devkit';
import { expectTypeScriptToCompile } from '../../utils/test/ts.spec';

describe('openApiTsClientGenerator - FastAPI', () => {
  let tree: Tree;
  const title = 'TestApi';

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  const validateTypeScript = (paths: string[]) => {
    expectTypeScriptToCompile(tree, paths);
  };

  it('should generate valid TypeScript for FastAPI echo endpoint with validation errors', async () => {
    const spec: Spec = {
      openapi: '3.1.0',
      info: { title, version: '0.1.0' },
      components: {
        schemas: {
          EchoOutput: {
            properties: {
              message: {
                title: 'Message',
                type: 'string',
              },
            },
            required: ['message'],
            title: 'EchoOutput',
            type: 'object',
          },
          HTTPValidationError: {
            properties: {
              detail: {
                items: {
                  $ref: '#/components/schemas/ValidationError',
                },
                title: 'Detail',
                type: 'array',
              },
            },
            title: 'HTTPValidationError',
            type: 'object',
          },
          InternalServerErrorDetails: {
            properties: {
              detail: {
                title: 'Detail',
                type: 'string',
              },
            },
            required: ['detail'],
            title: 'InternalServerErrorDetails',
            type: 'object',
          },
          ValidationError: {
            properties: {
              loc: {
                items: {
                  anyOf: [
                    {
                      type: 'string',
                    },
                    {
                      type: 'integer',
                    },
                  ],
                },
                title: 'Location',
                type: 'array',
              },
              msg: {
                title: 'Message',
                type: 'string',
              },
              type: {
                title: 'Error Type',
                type: 'string',
              },
            },
            required: ['loc', 'msg', 'type'],
            title: 'ValidationError',
            type: 'object',
          },
        },
      },
      paths: {
        '/echo': {
          get: {
            operationId: 'echo',
            parameters: [
              {
                in: 'query',
                name: 'message',
                required: true,
                schema: {
                  title: 'Message',
                  type: 'string',
                },
              },
            ],
            responses: {
              '200': {
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/EchoOutput',
                    },
                  },
                },
                description: 'Successful Response',
              },
              '422': {
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/HTTPValidationError',
                    },
                  },
                },
                description: 'Validation Error',
              },
              '500': {
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/InternalServerErrorDetails',
                    },
                  },
                },
                description: 'Internal Server Error',
              },
            },
            summary: 'Echo',
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    const mockFetch = vi.fn();

    // Test successful response
    mockFetch.mockResolvedValue({
      status: 200,
      json: vi.fn().mockResolvedValue({
        message: 'Hello, World',
      }),
    });

    const response = await callGeneratedClient(client, mockFetch, 'echo', {
      message: 'Hello, World',
    });

    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/echo?message=Hello%2C%20World`,
      expect.objectContaining({
        method: 'GET',
      }),
    );
    expect(response).toEqual({ message: 'Hello, World' });

    // Test validation error response (422)
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 422,
      json: vi.fn().mockResolvedValue({
        detail: [
          {
            loc: ['query', 'message'],
            msg: 'field required',
            type: 'value_error.missing',
          },
        ],
      }),
    });

    await expect(
      callGeneratedClient(client, mockFetch, 'echo', {}),
    ).rejects.toThrow();

    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/echo`,
      expect.objectContaining({
        method: 'GET',
      }),
    );

    // Test internal server error response (500)
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 500,
      json: vi.fn().mockResolvedValue({
        detail: 'Internal server error occurred',
      }),
    });

    await expect(
      callGeneratedClient(client, mockFetch, 'echo', { message: 'test' }),
    ).rejects.toThrow();

    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/echo?message=test`,
      expect.objectContaining({
        method: 'GET',
      }),
    );
  });

  it('should handle FastAPI validation error schemas with anyOf types', async () => {
    const spec: Spec = {
      openapi: '3.1.0',
      info: { title, version: '0.1.0' },
      components: {
        schemas: {
          ValidationError: {
            properties: {
              loc: {
                items: {
                  anyOf: [{ type: 'string' }, { type: 'integer' }],
                },
                title: 'Location',
                type: 'array',
              },
              msg: {
                title: 'Message',
                type: 'string',
              },
              type: {
                title: 'Error Type',
                type: 'string',
              },
            },
            required: ['loc', 'msg', 'type'],
            title: 'ValidationError',
            type: 'object',
          },
        },
      },
      paths: {
        '/test': {
          post: {
            operationId: 'testValidation',
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    properties: {
                      value: { type: 'integer' },
                    },
                    required: ['value'],
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: { type: 'string' },
                  },
                },
              },
              '422': {
                content: {
                  'application/json': {
                    schema: {
                      properties: {
                        detail: {
                          items: {
                            $ref: '#/components/schemas/ValidationError',
                          },
                          type: 'array',
                        },
                      },
                      type: 'object',
                    },
                  },
                },
                description: 'Validation Error',
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    const mockFetch = vi.fn();

    // Test successful request
    mockFetch.mockResolvedValue({
      status: 200,
      text: vi.fn().mockResolvedValue('success'),
    });

    const response = await callGeneratedClient(
      client,
      mockFetch,
      'testValidation',
      { value: 42 },
    );

    expect(mockFetch).toHaveBeenCalledWith(
      `${baseUrl}/test`,
      expect.objectContaining({
        method: 'POST',
        body: JSON.stringify({ value: 42 }),
      }),
    );
    expect(response).toBe('success');

    // Test validation error with mixed location types
    mockFetch.mockClear();
    mockFetch.mockResolvedValue({
      status: 422,
      json: vi.fn().mockResolvedValue({
        detail: [
          {
            loc: ['body', 'value'],
            msg: 'ensure this value is greater than 0',
            type: 'value_error.number.not_gt',
          },
          {
            loc: [0, 'nested_field'],
            msg: 'field required',
            type: 'value_error.missing',
          },
        ],
      }),
    });

    await expect(
      callGeneratedClient(client, mockFetch, 'testValidation', { value: -1 }),
    ).rejects.toThrow();
  });
});
