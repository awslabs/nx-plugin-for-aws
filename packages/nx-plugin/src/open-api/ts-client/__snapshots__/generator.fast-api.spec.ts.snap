// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openApiTsClientGenerator - FastAPI > should generate valid TypeScript for FastAPI echo endpoint with validation errors 1`] = `
"export type EchoOutput = {
  message: string;
};
export type EchoRequestQueryParameters = {
  message: string;
};
export type HTTPValidationError = {
  detail?: Array<ValidationError>;
};
export type InternalServerErrorDetails = {
  detail: string;
};
export type LocationItem = string | number;
export type ValidationError = {
  loc: Array<LocationItem>;
  msg: string;
  type: string;
};

export type EchoRequest = EchoRequestQueryParameters;
export type Echo422Error = {
  status: 422;
  error: HTTPValidationError;
};
export type Echo500Error = {
  status: 500;
  error: InternalServerErrorDetails;
};
export type EchoError = Echo422Error | Echo500Error;
"
`;

exports[`openApiTsClientGenerator - FastAPI > should generate valid TypeScript for FastAPI echo endpoint with validation errors 2`] = `
"import type {
  EchoOutput,
  EchoRequestQueryParameters,
  HTTPValidationError,
  InternalServerErrorDetails,
  LocationItem,
  ValidationError,
  EchoRequest,
} from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static EchoOutput = {
    toJson: (model: EchoOutput): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.message === undefined
          ? {}
          : {
              message: model.message,
            }),
      };
    },
    fromJson: (json: any): EchoOutput => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        message: json['message'],
      };
    },
  };

  public static EchoRequestQueryParameters = {
    toJson: (model: EchoRequestQueryParameters): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.message === undefined
          ? {}
          : {
              message: model.message,
            }),
      };
    },
    fromJson: (json: any): EchoRequestQueryParameters => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        message: json['message'],
      };
    },
  };

  public static HTTPValidationError = {
    toJson: (model: HTTPValidationError): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.detail === undefined
          ? {}
          : {
              detail: model.detail.map($IO.ValidationError.toJson),
            }),
      };
    },
    fromJson: (json: any): HTTPValidationError => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['detail'] === undefined
          ? {}
          : {
              detail: (json['detail'] as Array<any>).map(
                $IO.ValidationError.fromJson,
              ),
            }),
      };
    },
  };

  public static InternalServerErrorDetails = {
    toJson: (model: InternalServerErrorDetails): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.detail === undefined
          ? {}
          : {
              detail: model.detail,
            }),
      };
    },
    fromJson: (json: any): InternalServerErrorDetails => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        detail: json['detail'],
      };
    },
  };

  public static LocationItem = {
    toJson: (model: LocationItem): any => {
      if (model === undefined || model === null) {
        return model;
      }
      if (typeof model === 'string') {
        return model;
      }
      if (typeof model === 'number') {
        return model;
      }
      return model;
    },
    fromJson: (json: any): LocationItem => {
      if (json === undefined || json === null) {
        return json;
      }
      if (typeof json === 'string') {
        return json;
      }
      if (typeof json === 'number') {
        return json;
      }
      return json;
    },
  };

  public static ValidationError = {
    toJson: (model: ValidationError): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.loc === undefined
          ? {}
          : {
              loc: model.loc.map((item0) => item0),
            }),
        ...(model.msg === undefined
          ? {}
          : {
              msg: model.msg,
            }),
        ...(model.type === undefined
          ? {}
          : {
              type: model.type,
            }),
      };
    },
    fromJson: (json: any): ValidationError => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        loc: (json['loc'] as Array<any>).map((item0) => item0),
        msg: json['msg'],
        type: json['type'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.echo = this.echo.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async echo(input: EchoRequest): Promise<EchoOutput> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } =
      $IO.EchoRequestQueryParameters.toJson(input);
    const headerParameters: { [key: string]: any } = {};
    const collectionFormats = {
      message: 'multi',
    } as const;

    const body = undefined;

    const response = await this.$fetch(
      this.$url('/echo', pathParameters, queryParameters, collectionFormats),
      {
        headers: this.$headers(headerParameters, collectionFormats),
        method: 'GET',
        body,
      },
    );

    if (response.status === 200) {
      return $IO.EchoOutput.fromJson(await response.json());
    }
    if (response.status === 422) {
      throw {
        status: response.status,
        error: $IO.HTTPValidationError.fromJson(await response.json()),
      };
    }
    if (response.status === 500) {
      throw {
        status: response.status,
        error: $IO.InternalServerErrorDetails.fromJson(await response.json()),
      };
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }
}
"
`;

exports[`openApiTsClientGenerator - FastAPI > should handle FastAPI validation error schemas with anyOf types 1`] = `
"export type LocationItem = string | number;
export type TestValidation422Response = {
  detail?: Array<ValidationError>;
};
export type TestValidationRequestContent = {
  value: number;
};
export type ValidationError = {
  loc: Array<LocationItem>;
  msg: string;
  type: string;
};

export type TestValidationRequest = TestValidationRequestContent | undefined;
export type TestValidation422Error = {
  status: 422;
  error: TestValidation422Response;
};
export type TestValidationError = TestValidation422Error;
"
`;

exports[`openApiTsClientGenerator - FastAPI > should handle FastAPI validation error schemas with anyOf types 2`] = `
"import type {
  LocationItem,
  TestValidation422Response,
  TestValidationRequestContent,
  ValidationError,
  TestValidationRequest,
} from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static LocationItem = {
    toJson: (model: LocationItem): any => {
      if (model === undefined || model === null) {
        return model;
      }
      if (typeof model === 'string') {
        return model;
      }
      if (typeof model === 'number') {
        return model;
      }
      return model;
    },
    fromJson: (json: any): LocationItem => {
      if (json === undefined || json === null) {
        return json;
      }
      if (typeof json === 'string') {
        return json;
      }
      if (typeof json === 'number') {
        return json;
      }
      return json;
    },
  };

  public static TestValidation422Response = {
    toJson: (model: TestValidation422Response): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.detail === undefined
          ? {}
          : {
              detail: model.detail.map($IO.ValidationError.toJson),
            }),
      };
    },
    fromJson: (json: any): TestValidation422Response => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        ...(json['detail'] === undefined
          ? {}
          : {
              detail: (json['detail'] as Array<any>).map(
                $IO.ValidationError.fromJson,
              ),
            }),
      };
    },
  };

  public static TestValidationRequestContent = {
    toJson: (model: TestValidationRequestContent): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.value === undefined
          ? {}
          : {
              value: model.value,
            }),
      };
    },
    fromJson: (json: any): TestValidationRequestContent => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        value: json['value'],
      };
    },
  };

  public static ValidationError = {
    toJson: (model: ValidationError): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.loc === undefined
          ? {}
          : {
              loc: model.loc.map((item0) => item0),
            }),
        ...(model.msg === undefined
          ? {}
          : {
              msg: model.msg,
            }),
        ...(model.type === undefined
          ? {}
          : {
              type: model.type,
            }),
      };
    },
    fromJson: (json: any): ValidationError => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        loc: (json['loc'] as Array<any>).map((item0) => item0),
        msg: json['msg'],
        type: json['type'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.testValidation = this.testValidation.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async testValidation(input?: TestValidationRequest): Promise<string> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};
    if (!this.$config.options?.omitContentTypeHeader) {
      headerParameters['Content-Type'] = 'application/json';
    }
    const body =
      input === undefined
        ? undefined
        : typeof input === 'object'
          ? JSON.stringify($IO.TestValidationRequestContent.toJson(input))
          : String($IO.TestValidationRequestContent.toJson(input));

    const response = await this.$fetch(
      this.$url('/test', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'POST',
        body,
      },
    );

    if (response.status === 200) {
      return await response.text();
    }
    if (response.status === 422) {
      throw {
        status: response.status,
        error: $IO.TestValidation422Response.fromJson(await response.json()),
      };
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }
}
"
`;
